### DISCUSSION TITLE
Resources to become a engineer.

### DISCUSSION THREAD

**Discussion by MarkosComK** [community/community#138120] [Discover: GitHub Best Practices]
**Discussion Topic: Resources and Repositories for Becoming a Software Engineer**

Hi everyone! 👋

I'm currently on the path to becoming a software engineer and would love to know what resources, GitHub repositories, and guidelines you recommend for someone aspiring to enter the field.

What are some of the best open-source projects, coding challenges, or practice repositories for building foundational skills? Additionally, are there any specific resources or learning paths (e.g., guides, blogs, YouTube channels) that helped you on your journey to becoming a software engineer?

Would also love to hear any advice on topics like:

- Essential programming languages and frameworks to focus on
- Best practices for version control and project management
- Key concepts like algorithms, data structures, and system design
- Tips for contributing to open-source projects

Looking forward to hearing your suggestions and learning from your experiences!

**Comment by SirAlabar** [comment]
It's great that you're pursuing a career in software engineering! Here are some resources and tips that can help you build your foundational skills:
1. GitHub Repositories & Open-Source Projects

    [FreeCodeCamp](https://github.com/freeCodeCamp/freeCodeCamp): A fantastic open-source project with coding challenges and certifications in web development, algorithms, and data structures.
    [The Algorithms](https://github.com/TheAlgorithms): A collection of algorithm implementations in multiple programming languages.
    [Awesome Lists](https://github.com/sindresorhus/awesome): A curated list of awesome frameworks, libraries, software, and more.
    [Real-World App Repos](https://github.com/gothinkster/realworld): Full-stack apps built with various languages and frameworks that are great for understanding how to structure projects.

2. Coding Challenges & Practice

    [LeetCode](https://leetcode.com/): Excellent for practicing algorithms and preparing for technical interviews.
    [HackerRank](https://www.hackerrank.com/): Focuses on competitive programming and skill-building challenges.
    [Exercism](https://exercism.io/): Offers exercises in a variety of languages with mentoring from experienced developers.
    [Codewars](https://www.codewars.com/): A platform with challenges to improve your problem-solving skills.

3. Learning Resources

    YouTube Channels:
        [Traversy Media](https://www.youtube.com/c/TraversyMedia): For web development and programming tutorials.
        [CS50 by Harvard](https://www.youtube.com/c/cs50): The legendary intro to computer science.
        [Fireship](https://www.youtube.com/c/Fireship): Quick, to-the-point tutorials on programming and frameworks.
    Guides & Blogs:
        [Roadmap.sh](https://roadmap.sh/): Comprehensive guides for developers, including front-end, back-end, and DevOps.
        [Codecademy](https://www.codecademy.com/): Great for learning various languages and frameworks interactively.
        [MDN Web Docs](https://developer.mozilla.org/en-US/): The best place for learning HTML, CSS, and JavaScript.
    Books:
        “Clean Code” by Robert C. Martin: Essential reading for writing readable, maintainable code.
        “Introduction to Algorithms” by Cormen, Leiserson, Rivest, and Stein: A foundational book for understanding algorithms and data structures.

4. Programming Languages and Frameworks

    Languages:
        Python: For scripting, automation, and data science.
        JavaScript: Essential for web development (React, Node.js).
        Java/C++: Important for system-level programming and back-end development.
        SQL: Learn databases (MySQL, PostgreSQL).
    Frameworks:
        React: A must for front-end web development.
        Django/Flask: For Python-based back-end development.
        Spring Boot: For Java-based back-end development.

5. Version Control & Project Management

    Git: Mastering Git is crucial for version control. Explore branching, merging, rebasing, and GitHub workflows (pull requests, code reviews).
    Agile & Scrum: Learn project management methodologies like Agile and Scrum to understand team-based development and iterative progress.
    Trello/Jira: Tools for project and task management.

6. Key Concepts

    Algorithms & Data Structures: Make sure to have a solid understanding of sorting algorithms (merge sort, quicksort), data structures (arrays, linked lists, trees, graphs), and dynamic programming.
    System Design: Once you're comfortable with coding, learn system design principles (load balancing, databases, scaling) for designing large-scale applications.

7. Contributing to Open Source

    Start with beginner-friendly issues in repositories labeled "good first issue" or "help wanted".
    Learn how to fork projects, submit pull requests, and participate in discussions.
    Read the contribution guidelines of each project and start with documentation or bug fixes to get familiar with the codebase.

Additional Tips:

    Networking: Join developer communities (GitHub Discussions, Stack Overflow, Dev.to) and participate in coding events (hackathons).
    Build Projects: Apply what you learn by building real-world projects, such as personal portfolios, blogs, or even small games.

Good luck on your journey! Feel free to ask for more advice if needed! 😄
